### ISSUE TITLE
Support the new SPIFFE Auth validator in Envoy

### ISSUE THREAD

**Issue by azdagron** [spiffe/spire#2420]
Envoy has recently gained support for a SPIFFE Auth extension that supports authenticating identities from multiple trust domains with a single validation context.

It fits into the `custom_validator_config` field of the CertificateValidationContext.

This issue tracks the work to add support for this extension to the SDS implementation provided by the agent.

More info on the extension can be found [here](https://www.envoyproxy.io/docs/envoy/v1.18.2/api-v3/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto)

**Comment by MarcosDY** [comment]
In current implementation we allow returning a single Validation context with its bundle, based on the provided name:
```
            validation_context_sds_secret_config:
              name: "spiffe://domain.test"
              sds_config:
                api_config_source:
                  api_type: GRPC
                  grpc_services:
                    envoy_grpc:
                      cluster_name: spire_agent
                  transport_api_version: V3
                resource_api_version: V3
```
In this example we choose to return the bundle with TD `spiffe://domain.test`
Or if [default_bundle_name](https://github.com/spiffe/spire/blob/main/doc/spire_agent.md#sds-configuration) is used, it can be returned (it is done this way to be able to return the agent's bundle, for ISTIO compatibility).
It is even possible to return a federated bundle if it is chosen as a name.
However, with this new feature we must be able to return multiple bundles, that can be used to authenticate the caller's identity.
But when must we return all those new bundles?
Here we have some options:
1- add a new configurable in agent's level that enables this feature.
2- add a constant that a user can choose as `resource name` and when it is provided return a `custom_validation_context` instead of regular validation_context.
3- update default bundle implementation to start returning all.

**Comment by azdagron** [comment]
I don't think option 3 is safe to do. An deployment that upgraded would suddenly start federating on existing endpoints that used the default resource name.... this may or may not be a welcome change and certainly could be problematic if authorization policies didn't exist that took federated trust domains into account.

I think a combination of 1 & 2 could be possible. The configurable for the default resource name for the bundle is `default_bundle_name` and defaults to `ROOTCA`. We could add another configurable named something like `default_all_bundles_name` which defaulted to something like `ALL`. It is possible that somebody has set their `default_bundle_name` to our chosen default, but we could just log a warning and fall back to old behavior (i.e. only return the bundle for the trust domain) when that is the case.

So TL;DR:

| `default_bundle_name` | `default_all_bundles_name` | resource requested | bundle in `trusted_ca` | bundles in `spiffe` extension | warning |
| --- | --- | --- | --- | --- | --- |
|  X  |  Y  |  X  |  primary | primary bundle |
|  X  |  Y  |  Y  |  primary  | all |
|  X  |  X  |  X  |  primary | primary bundle | Y
| X | Y | `domain.test` | `domain.test` | `domain.test` |

**Comment by azdagron** [comment]
A small clarification. I think ideally the agent fails to start if both `default_bundle_name` and `default_all_bundles_name` are identical. We can't do that for back compat reasons (i.e.`default_bundle_name` is set to the default value we choose for `default_all_bundles_name`). I do think that the warning message that we log could indicate that this type of misconfiguration will, at a future point, cause the agent not to start. Then in 1.2.0 we can change the behavior to not allow the misconfiguration.

**Comment by MarcosDY** [comment]
I found that when using custom_validator_config on envoy below 1.18.0,  an error is returned
```
[34mupstream-proxy_1         |[0m time="2021-07-29T17:06:08Z" level=error msg="Envoy reported errors applying secrets" error="Protobuf message (type envoy.extensions.transport_sockets.tls.v3.CertificateValidationContext with unknown field set {12}) has unknown fields" method=StreamSecrets pid=9 resource_names="[spiffe://domain.test]" service=SDS.v3 subsystem_name=endpoints
```
custom_validator_config was added on 1.18 (https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.18.0)

I didnt found a way to determinate what version is the Envoy Proxy that is calling SDS, and even if we can determinate what version is calling us, it can result that we'll be adding more special cases in the future and that is risky.

So, an option is to add a bool `sds { allow_custom_validation_config = true  }`
When set it will add custom_validation_config as specified in previous comment.
And when `default_all_bundles_name` is requested but `allow_custom_validation_config` is false an error will be retuned

Summary: 
| allow | default_bundle_name | default_all_bundles_name | resource requested | bundle in `trusted_ca` | `bundles in spiffe extension`  | error |
|--| -- | -- | - | - | - | - |
| T | X | Y | X | primary | primary bundle |  
| T | X | Y | Y | primary | all |  
| T | X | X | X | primary | primary bundle | 
| T | X | Y | domain.test | domain.test | domain.test
| F | X | Y | X | primary | nil |  
| F | X | Y | Y | nil | nil |  Y |
| F | X | X | X | primary | nil  | 
| F | X | Y | domain.test | domain.test | nil  |

**Comment by MarcosDY** [comment]
Thinking it more, I figure out that instead of `allow_spiffe_validation` we can have a `use_spiffe_validator` that will replace the old  `trusted_ca` since `trusted_ca` is ignored if spiffe validator is set, and in older version is not possible to use `spiffe validation`

**Comment by MarcosDY** [comment]
good news, SDS provides envoy builded version inside node, so a configuration is not required.

https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-msg-config-core-v3-buildversion

I verified against 1.17:  `user_agent_name:\"envoy\"  user_agent_build_version:{version:{major_number:1  minor_number:17}`

**Comment by azdagron** [comment]
That is great news!
